src/sdk/object_sdk.js

```diff
@@ -64,6 +65,41 @@ class ObjectSDK {
             this.bucketspace_fs = new BucketSpaceFS({ fs_root: process.env.BUCKETSPACE_FS });
         }
         this.requesting_account = undefined;
+        this.abort_controller = new AbortController();
+    }
+
+    /**
+     * setup_abort_controller adds event handlers to the http request and response,
+     * in order to handle aborting requests gracefully. The `abort_controller` member will
+     * be used to signal async flows that abort was detected.
+     * @see {@link https://nodejs.org/docs/latest/api/globals.html#class-abortcontroller}
+     * @param {import('http').IncomingMessage} req 
+     * @param {import('http').ServerResponse} res 
+     */
+    setup_abort_controller(req, res) {
+        res.once('error', err => {
+            dbg.log0('response error:', err, req.url);
+            this.abort_controller.abort();
+        });
+
+        res.once('close', () => {
+            dbg.log0('response closed', req.url);
+            this.abort_controller.abort();
+        });
+
+        req.once('aborted', () => {
+            dbg.log0('request aborted', req.url);
+            this.abort_controller.abort();
+        });
+    }
+
+    throw_if_aborted() {
+        if (this.abort_controller.signal.aborted) throw new Error('request aborted signal');
+    }
+
+    add_abort_handler(handler) {
+        if (this.abort_controller.signal.aborted) setImmediate(handler);
+        return this.abort_controller.signal.addEventListener('abort', handler, { once: true });
     }
```

Comment:
```suggestion
    add_abort_handler(handler) {
        const s = this.abort_controller.signal;
        if (s.aborted) {
            setImmediate(handler);
        } else {
            s.addEventListener('abort', handler, { once: true });
        }
    }
```

Reply:
This suggested change makes sure we don't add the listener when aborted ^ as we don't want both setImmediate and addEventListener to be applied in this case.